### ISSUE TITLE
Backwards compatibility with older versions of macOS

### ISSUE THREAD

**Issue by thebigh2014** [EndlessSkyCommunity/ESLauncher2#64]
I'm not sure if this is possible but having the ESLauncher2 be compatible with older versions of macOS (I'm running 10.11) would be much appreciated.  My computer does meet all the other requirements so unless there's something in the code or dependencies that precludes my system I don't see this being a problem.

Or maybe at least if I had the ability to easily build it myself that would be helpful as no Xcode project is provided for a macOS build (so I'm not even sure how you build it).

**Comment by MCOfficer** [comment]
The restriction is the github CI that builds all packages, it can only run on catalina. We already set a flag to build for versions as old as 10.14, but i'm not sure if we can go older:
https://github.com/EndlessSkyCommunity/ESLauncher2/blob/e8592be279a29cdeb218b15462549412cb685df7/.github/workflows/build.yaml#L70

In the meantime, you can always compile ESL2 from source:
- Install the toolchain via [rustup](https://rustup.rs/) (default settings should suffice)
- in the ESL2 directory, run these commands: https://github.com/EndlessSkyCommunity/ESLauncher2/blob/e8592be279a29cdeb218b15462549412cb685df7/.github/workflows/build.yaml#L72-L84
- find ESL2 in `target/release/bundle/osx`

**Comment by tehhowch** [comment]
The catalina ci issue will probably affect your ability to use the launcher to play the continuous build as well.

I'm curious if we can implement a custom workflow for the macOS build to produce a data-only artifact and several different binaries, for target macOS versions, and then in this launcher, combine the data with the binary appropriate to the host system.

This would likely require experimentation on @thebigh2014 's part to see if we can change the XCode project target version and have the CI output respect that.

**Comment by thebigh2014** [comment]
@MCOfficer Generally speaking, Xcode allows targeting several OS versions back along with current and the next version.  For example in Xcode 8.2.1, the last version for 10.11, deployment targets range from 10.6 up to 10.12  If the same range is available for Xcode 11.x in 10.15, then supported deployment targets should be from 10.10 up to 11.

**Comment by tehhowch** [comment]
@thebigh2014 to test this myself would require a Mac. Short of being gifted one for development, that won't ever happen, so if changes to the XCode project to support older versions of macOS are possible, someone else will need to bed the one who makes those changes.

**Comment by thebigh2014** [comment]
@tehhowch The Xcode project file (project.pbxproj inside EndlessSky.xcodeproj) is a text file that can be easily edited.  The deployment targets are specified on lines 1125 (debug) and 1155 (release).  I could submit a PR with the changes if you'd like.

**Comment by tehhowch** [comment]
>   I could submit a PR with the changes if you'd like.

That is part of what I was suggesting above. I'm well aware of how to change what the ci does, but I cannot test that the changes have their intended effect:

> This would likely require experimentation on @thebigh2014 's part to see if we can change the XCode project target version and have the CI output respect that.

So in your fork of ES, you would make those changes and then verify that the artifact produced by the GitHub CI is playable on your machine. If it is, then we need to work with MCO to make ES Launcher capable of detecting (or told) which OSX binary to use on a per-user basis. Then we can update the main repo to build several macOS target versions, including one which works on your machine.

**Comment by thebigh2014** [comment]
Right, makes sense.  And the line numbers I quoted were from my local version, which was modified by my uses of Xcode.  I will see about trying to make the change in my fork with the GitHub CI.

**Comment by MCOfficer** [comment]
I'm starting to understand why game developers don't want to support niche OSes - So much work for such a small subset of users!

Just to give you a statement: ESL2 will not support MacOS versions that ES itself doesn't support. So in order to advance this issue, the continuous build needs to be more backwards compatible.

**Comment by tehhowch** [comment]
Also fair. @thebigh2014 you'd still do local testing on your fork to see if the CI can produce a binary you are able to use, and then we would move to the PR stage and ask Mac users of ES Launcher to test that PR instance.

We should move future conversation to https://github.com/endless-sky/endless-sky/issues/5013

(That issue also highlights that the XCode target macOS version is not the only obstacle to be overcome.)

**Comment by thebigh2014** [comment]
@MCOfficer Hence why I merely asked if it was possible and why I am willing to volunteer my time to see if it can be done or not.  And I agree that ESL2 should only support macOS versions that ES supports as well.

@tehhowch Right, that would be the issue to continue this discussion at.  Btw., just to let you know, I have not had any success thus far.  I even tried using Xcode 10.3, which is available on the Github CI, but no luck as of yet.  There are some things left to try but in the end it may be up to individuals, like me, to manually create and upload unofficial builds for older OSes.

Oh and the error I'm getting is the same as before (as @MCOfficer mentioned in https://github.com/endless-sky/endless-sky/issues/5013#issuecomment-619352506 )

**Comment by MCOfficer** [comment]
I meant no offense, just wanted to clarify. I just updated endless-sky/endless-sky#5013 because apparently i'd forgotten it.

**Comment by thebigh2014** [comment]
No offense was taken.  I totally understand your point of view as I'm a software developer dealing with backwards compatibility and breaking changes with web applications.

**Comment by TonyKZ1** [comment]
@thebigh2014 Was you able to get the ESLauncher2 to work on macOS 10.13?
